RestaurantReservation.API/Controllers/ReservationController.cs

```diff
@@ -0,0 +1,193 @@
+        /// <response code="422">There's an invalid foreign key returned in the response body.</response>
+        [HttpPost]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
+        [ProducesResponseType(typeof(ReservationDTO), StatusCodes.Status201Created)]
+        public async Task<IActionResult> CreateReservationAsync(
+            ReservationWithoutIdDTO newReservation)
+        {
+            var validationResult = await _validator.ValidateAsync(newReservation);
+
+            if (!validationResult.IsValid)
+                return BadRequest(validationResult.GetImportantProperties());
+
+            int newId;
+
+            try
+            {
+                newId = await _reservationService.
+                CreateAsync(_mapper.Map<Reservation>(newReservation));
+            }
+            catch (DbUpdateException ex)
+            {
+                return UnprocessableEntity(
+                    $"Invalid foreign key: {ex.ExtractForeignKey()}");
+            }
+
+            var responseReservation = _mapper.Map<ReservationDTO>(newReservation);
+            responseReservation.Id = newId;
+
+            return Created($"api/reservations/{newId}", responseReservation);
+        }
+
+        /// <summary>
+        /// Update a reservation with a specific Id.
+        /// </summary>
+        /// <param name="reservationId">The Id of the reservation to update.</param>
+        /// <param name="updatedReservation">The reservation with updated value(s).</param>
+        /// <response code="404">The reservation with the given Id doesn't exist.</response>
+        /// <response code="422">There's an invalid foreign key returned in the response body.</response>
+        /// <response code="204">The reservation is updated successfully.</response>
+        [HttpPut("{reservationId}")]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        public async Task<IActionResult> UpdateReservationAsync(
+            int reservationId,
+            ReservationWithoutIdDTO updatedReservation)
+        {
+            var validationResult = await _validator.ValidateAsync(updatedReservation);
+
+            if (!validationResult.IsValid)
+                return BadRequest(validationResult.GetImportantProperties());
+
+            var reservationWithId = _mapper.Map<Reservation>(updatedReservation);
+            reservationWithId.Id = reservationId;
+
+            try
+            {
+                await _reservationService.UpdateAsync(reservationWithId);
+            }
+            catch (KeyNotFoundException)
+            {
+                return NotFound();
+            }
+            catch (DbUpdateException ex)
```

Comment:
we can not guarantee we would have only these exceptions
we have to add a catch for a general exception at the end